devito/passes/clusters/aliases.py

```diff
@@ -81,244 +81,303 @@ def cire(clusters, mode, sregistry, options, platform):
-        # Forbid CIRE involving Dimension-independent dependencies, e.g.:
-        # r0 = ...
-        # u[x, y] = ... r0*a[x, y] ...
-        # NOTE: if one uses the DSL in a conventional way and sticks to the default
-        # compilation pipelines where CSE always happens after CIRE, then `exclude`
-        # will always be empty
-        exclude = {i.source.indexed for i in context[None].scope.d_flow.independent()}
+    def process(self, clusters):
+        processed = []
+        for c in clusters:
+            if not c.is_dense:
+                processed.append(c)
+                continue
 
-        mapper = Uxmapper()
-        for e in exprs:
-            for i in search_potential_deriv(e, n):
-                if i.free_symbols & exclude:
-                    continue
+            # Rule out Dimension-independent dependencies, e.g.:
+            # r0 = ...
+            # u[x, y] = ... r0*a[x, y] ...
+            exclude = {i.source.indexed for i in c.scope.d_flow.independent()}
 
-                key = lambda a: a.is_Add
-                terms, others = split(i.args, key)
-
-                if self._opt_maxalias:
-                    # Treat `e` as an FD expression and pull out the derivative
-                    # coefficient from `i`
-                    # Note: typically derivative coefficients are numbers, but
-                    # sometimes they could be provided in symbolic form through an
-                    # arbitrary Function.  In the latter case, we rely on the
-                    # heuristic that such Function's basically never span the whole
-                    # grid, but rather a single Grid dimension (e.g., `c[z, n]` for a
-                    # stencil of diameter `n` along `z`)
-                    if e.grid is not None and terms:
-                        key = partial(maybe_coeff_key, e.grid)
-                        others, more_terms = split(others, key)
-                        terms += more_terms
-
-                mapper.add(i, self._make_symbol, terms)
+            # TODO: to process third- and higher-order derivatives, we could
+            # extend this by calling `_aliases_from_clusters` repeatedly until
+            # `made` is empty. To be investigated
+            made = self._aliases_from_clusters([c], exclude, self._lookup_key(c))
 
-        return mapper
+            processed.extend(flatten(made) or [c])
 
-    def _in_writeto(self, dim, cluster):
-        return self._opt_maxpar and PARALLEL in cluster.properties[dim]
+        return processed
 
-    def _selector(self, e, naliases):
-        if naliases <= 1:
-            return 0
-        else:
-            return estimate_cost(e, True)*naliases // self._opt_mincost
+    def _generate(self, exprs, exclude):
+        # E.g., extract `u.dx*a*b` and `u.dx*a*c` from `[(u.dx*a*b).dy`, `(u.dx*a*c).dy]`
+        def cbk_search(expr):
+            if isinstance(expr, EvalDerivative) and not expr.base.is_Function:
```

Comment:
Note for reviewers: this is the game changer -- we can finally look for `EvalDerivative`, rather than speculatively searching all sort of expressions in sum-of-product form